### ISSUE TITLE
[MSOURCES-144] Include sources for multi-release jars

### ISSUE THREAD

**Issue by jira-importer** [apache/maven-source-plugin#215]
**[Tagir Valeev](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=lany)** opened **[MSOURCES-144](https://issues.apache.org/jira/browse/MSOURCES-144?redirect=false)** and commented

Java 9 introduced multi-release JARs seven years ago (see [JEP 238](https://openjdk.org/jeps/238)). Since then, Maven support for multi-release JARs was gradually improved, and now it's quite easy to create and deploy such a project. However, I see a significant unresolved problem: the source files for alternative versions of class-files are not automatically added to the source jar, and I see no easy way to fix this.

Consider the attached example. It contains two versions of com.example.DefaultVersion class, one for Java 8, and one for Java 9. Execute `mvn verify`. The resulting multi-release-sample-1.0.jar properly contains com/example/DefaultVersion.class as Java 8 compiled version and META-INF/versions/9/com/example/DefaultVersion.class as Java 9 compiled version. However, the source JAR multi-release-sample-1.0-sources.jar contains no Java 9-specific version. I expect that the source jar should add Java 9 source version under the path META-INF/versions/9/com/example/DefaultVersion.java

This has an unpleasant effect: the library sources are incomplete, so the library user cannot easily browse them without using decompiler or looking for the source repository. This complicates navigation and debugging in IDEs. In fact, I've checked dozens of multi-release JARs in Maven Central, and none of them contain version-specific sources. It would be great to solve this on the maven-source-plugin side, so the library maintainers need only to bump source-plugin version to publish the sources properly.

Thank you in advance,
Tagir Valeev.

---

**Affects:** 3.3.0

**Attachments:**
- [multirelease.zip](https://issues.apache.org/jira/secure/attachment/13067152/multirelease.zip) (_2.59 kB_)

1 votes, 3 watchers

**Comment by amaembo** [comment]
@bmarwell thank you for looking into this. IntelliJ IDEA is not opinionated about the directory layout of the project. One may register any directory as a source root. If the configuration comes from Maven, we try our best to import the project model as close as possible to the original build system, so we will follow any convention/defaults supported by Maven if the project is Maven-based.

As for source.jar layout, I agree that we require a convention here about the location of the sources inside META-INF/versions/X. Here, I don't see any space for discussion. Unless I'm missing something, there's no better alternative to this layout, so we will happily follow it. Moreover, I believe this already works in IntelliJ IDEA, though it's not well-tested, as so far I haven't seen any source jar with such a layout in the Central.

**Comment by laeubi** [comment]
> All IDEs need to agree on this (IDEA, NetBeans, VSCode and Eclipse), so I will put it on the mailing list. 

I don't think so, an IDE should support the user as much as possible, so likely want to adapt to different conventions.

Anyways MR is special because the same **name** maps to different things in the output so one can not put it in one folder, but even if one would simply copy each source folder in one folder in the source jar it would be doable. PDE supports different layouts e.g. OSGi with `OSGI-OPT` folder, maven source jars, Eclipse specific source infos that even allows subfolders by a Manifest header and so on.

In any case I think using the same layout for a "flat" source is likely a good choice and I don't see anything that speaks against this as it follows the same semantic rules like defined for the binary artifact, so why make it more complicated?

**Comment by bmarwell** [comment]
I haven't updated this issue in a while. Maven 4 will make this possible, but let me check whether this is possible in Maven 3 with some tricks.
